File: Dynamic Programming/198. House Robber/198. House Robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]

        for i in range(1, len(nums)):
            if i == 1:
                nums[i] = max(nums[i], nums[i - 1])
            else:
                nums[i] = max(nums[i - 1], nums[i - 2] + nums[i])
        
        return nums[-1]

# Time: O(n)
# Space: O(1)

File: Dynamic Programming/198. House Robber/198. House Robber.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function (nums) {
  if (nums.length === 1) return nums[0];

  for (let i = 1; i < nums.length; i++) {
    if (i === 1) nums[i] = Math.max(nums[i], nums[i - 1]);
    else nums[i] = Math.max(nums[i - 1], nums[i - 2] + nums[i]);
  }

  return nums.at(-1);
};
// Time: O(n)
// Space: O(1)
